features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,414 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef PCL_FEATURES_IMPL_GASD_H_
+#define PCL_FEATURES_IMPL_GASD_H_
+
+#include <pcl/features/gasd.h>
+#include <pcl/common/transforms.h>
+#include <pcl/point_types_conversion.h>
+
+#include <vector>
+
+namespace pcl
+{
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::compute (PointCloudOut &output)
+  {
+    if (!Feature<PointInT, PointOutT>::initCompute ())
+    {
+      output.width = output.height = 0;
+      output.points.clear ();
+      return;
+    }
+
+    // Resize the output dataset
+    output.width = output.height = 1;
+    output.points.resize (1);
+
+    // Copy header and is_dense flag from input
+    output.header = surface_->header;
+    output.is_dense = surface_->is_dense;
+
+    // Perform the actual feature computation
+    computeFeature (output);
+
+    Feature<PointInT, PointOutT>::deinitCompute ();
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDEstimation<PointInT, PointOutT>::computeAlignmentTransform ()
+  {
+    Eigen::Vector4f centroid;
+    Eigen::Matrix3f covariance_matrix;
+
+    // compute centroid of the object's partial view
+    pcl::compute3DCentroid (*surface_, *indices_, centroid);
+
+    // compute covariance matrix from points and centroid of the object's partial view
+    pcl::computeCovarianceMatrix (*surface_, *indices_, centroid, covariance_matrix);
+
+    Eigen::Matrix3f eigenvectors;
+    Eigen::Vector3f eigenvalues;
+
+    // compute eigenvalues and eigenvectors of the covariance matrix
+    pcl::eigen33 (covariance_matrix, eigenvectors, eigenvalues);
+
+    // z axis of the reference frame is the eigenvector associated with the minimal eigenvalue
+    Eigen::Vector3f z_axis = eigenvectors.col (0);
+
+    // if angle between z axis and viewing direction is in the [-90 deg, 90 deg] range, then z axis is negated
+    if (z_axis.dot (view_direction_) > 0)
+    {
+      z_axis = z_axis * -1;
+    }
+
+    // x axis of the reference frame is the eigenvector associated with the maximal eigenvalue 
+    const Eigen::Vector3f x_axis = eigenvectors.col (2);
+
+    // y axis is the cross product of z axis and x axis
+    const Eigen::Vector3f y_axis = z_axis.cross (x_axis);
+
+    const Eigen::Vector3f centroid_xyz = centroid.head<3> ();
+
+    // compute alignment transform from axes and centroid
+    transform_ << x_axis[0], x_axis[1], x_axis[2], -x_axis.dot (centroid_xyz),
+                  y_axis[0], y_axis[1], y_axis[2], -y_axis.dot (centroid_xyz),
+                  z_axis[0], z_axis[1], z_axis[2], -z_axis.dot (centroid_xyz),
+                  0.0f, 0.0f, 0.0f, 1.0f;
```

Comment:
With this operator you can actually write 

transform_  = x_axis, -x_axis.dot(...), y_axis, -y_axis.dot(...), ...